vmsyscall/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! syscall serialize/deserialize
+//!
+//! Currently it uses a hard coded page and an I/O trigger.
+//! We might want to switch to MMIO.
+
+#![deny(missing_docs)]
+#![deny(clippy::all)]
+#![no_std]
+
+/// The IO trigger port for vm syscall proxy
+pub const TRIGGER_PORT: u16 = 0xFF;
+
+use serde::{Deserialize, Serialize};
+
+/// The syscalls to be serialized/deserialized via serde
+/// for the Hypervisor <-> VM syscall proxy
+#[derive(Serialize, Deserialize, Debug)]
+pub enum VmSyscall {
+    /// int madvise(void *addr, size_t length, int advice);
+    Madvise {
+        /// see madvise(2)
+        addr: usize,
+        /// see madvise(2)
+        length: usize,
+        /// see madvise(2)
+        advice: i32,
+    },
+    /// void *mmap(void *addr, size_t length, int prot, int flags, …);
+    Mmap {
+        /// see mmap(2)
+        addr: usize,
+        /// see mmap(2)
+        length: usize,
+        /// see mmap(2)
+        prot: i32,
+        /// see mmap(2)
+        flags: i32,
+    },
+    /// void *mremap(void *old_address, size_t old_size, size_t new_size, int flags, ... /* void *new_address */);
+    Mremap {
+        /// see mremap(2)
+        old_address: usize,
+        /// see mremap(2)
+        old_size: usize,
+        /// see mremap(2)
+        new_size: usize,
+        /// see mremap(2)
+        flags: i32,
+    },
+    /// int munmap(void *addr, size_t length);
+    Munmap {
+        /// see munmap(2)
+        addr: usize,
+        /// see munmap(2)
+        length: usize,
+    },
+    /// int mprotect(void *addr, size_t len, int prot);
+    Mprotect {
+        /// see mprotect(2)
+        addr: usize,
+        /// see mprotect(2)
+        length: usize,
+        /// see mprotect(2)
+        prot: i32,
+    },
+    // Todo: extend with needed hypervisor proxy syscalls
```

Comment:
Who do we find out which syscalls are needed? Are these the ones required by the runtime which is added in #139?

Reply:
For the sev kernel case we only proxy those, which the kernel can't handle itsself.. like network related syscalls like socket(2). And if the kernel does not want to do extensive memory handling, we want to proxy the mmap family, too.

Reply:
currently this is just a skeleton crate, which will be extended and stripped a lot in the future.